### ISSUE TITLE
UpdatableRecord::merge doesn't work in the presence of a unique constraint and in the absence of a primary key value

### ISSUE THREAD

**Issue by AnghelLeonard** [jOOQ/jOOQ#11552]
### Expected behavior
A working MySQL `ON DUPLICATE KEY UPDATE`.

### Actual behavior
A MySQL `ON DUPLICATE KEY UPDATE` that does do anything.

### Steps to reproduce the problem
Consider this table (`BANK_TRANSACTION`):
![image](https://user-images.githubusercontent.com/3829643/109685086-9118f500-7b89-11eb-8f2f-9efe67ddb6f9.png)

The `uin` is unique:

`CONSTRAINT unique_uin UNIQUE (uin)`

Next, I create a new `BankTransactionRecord` without setting the primary key (`TRANSACTION_ID` is auto-incremented and is `null` in this `BankTransactionRecord`), but setting `UID` as the existent `KMO1677VC45` - third row.

Calling `merge()` on this `Record` results in this SQL:

```sql
insert into bank_transaction (
  bank_name, bank_iban, transfer_amount, 
  customer_number, check_number,  uin
) 
values 
  (
    'Some Bank', 'IBAN2', 53343, 
    447, 'AO757239', 'KMO1677VC45'
  ) on duplicate key 
update 
  bank_name = case when transaction_id is null then 'Some Bank' else bank_name end, 
  bank_iban = case when transaction_id is null then 'IBAN2' else bank_iban end, 
  ...
  uin = case when transaction_id is null then 'KMO1677VC45' else uin end
```

This query doesn't do anything since the condition `transaction_id is null then ...` is never true. The `KMO1677VC45` transaction ID is 3. If I manually add `is not null` it works. 

### Versions

- jOOQ:3.14.8
- Java:14
- Database (include vendor):MySQL
- OS:-
- JDBC Driver (include name if inofficial driver):-

**Comment by lukaseder** [comment]
Looking into this now. Somewhat related: https://github.com/jOOQ/jOOQ/issues/13170

**Comment by lukaseder** [comment]
I can reproduce the problem with this table:

```sql
CREATE TABLE t_11552 (
  id INTEGER AUTO_INCREMENT,
  cd INTEGER UNIQUE,
  xy INTEGER,
  
  CONSTRAINT t_11552_pk PRIMARY KEY (id)
);
```

And this test case:

```sql
public void testUpdatablesMergeWithIdentityAndUniqueKey() throws Exception {
    assumeTableExists(T_11552);
    DSLContext ctx = new ImplicitJoinTests<>(delegate).create();

    try {
        T_11552Record r1 = ctx.newRecord(T_11552);
        r1.setCd(1);
        r1.setXy(1);
        r1.merge();

        T_11552Record r2 = ctx.fetchOne(T_11552);
        assertEquals(r1.getCd(), r2.getCd());
        assertEquals(r1.getXy(), r2.getXy());

        T_11552Record r3 = ctx.newRecord(T_11552);
        r3.setCd(1);
        r3.setXy(2);
        r3.merge();

        T_11552Record r4 = ctx.fetchOne(T_11552);
        assertEquals(r3.getCd(), r4.getCd());
        assertEquals(r3.getXy(), r4.getXy());
        assertEquals(r2.getId(), r4.getId());
    }
    finally {
        ignoreThrows(() -> ctx.deleteFrom(T_11552).execute());
    }
}
```

The generated SQL on MySQL is:

```sql
insert into t_11552 (CD, XY)
values (
  1, 
  2
)
on duplicate key update
  t_11552.CD = case
    when t_11552.ID is null then 1
    else t_11552.CD
  end,
  t_11552.XY = case
    when t_11552.ID is null then 2
    else t_11552.XY
  end
```

**Comment by lukaseder** [comment]
The problem also happens with `MERGE` based implementations of `UpdatableRecord::merge`, e.g. in H2:

```sql
select ID
from final table (
  merge into T_11552
  using (
    select CD, XY
    from (
      select 1, 2
    ) t (CD, XY)
  ) t
  on (
    T_11552.ID = null -- This should be replaced by IS NULL, but is mostly cosmetic
    or T_11552.CD = t.CD
  )
  when matched and T_11552.ID is null then update set -- This check is wrong
    T_11552.CD = 1,
    T_11552.XY = 2
  when not matched then insert (CD, XY)
  values (
    t.CD, 
    t.XY
  )
) T_11552
```

**Comment by lukaseder** [comment]
The main problem here is that we're not considering the `changed()` flag on the primary key value, meaning that the user has not set the value explicitly. This is especially true when we have a candidate key (unique), which acts as a non-primary natural key, and a surrogate primary key, which the client might not know. The client will then never set the `ID` value explicitly, as they're always working with the record via the natural key, instead.

**Comment by lukaseder** [comment]
Looking at the call stack of the `Tools::condition` method, which is responsible for the `IS NULL` predicate:

![image](https://user-images.githubusercontent.com/734593/157676499-31a08285-e25b-43fc-b0b6-c89c8fda1bd3.png)

It becomes clear that changing things in this area is very delicate. If the key value is not available, we mustn't simply ignore it, especially when running `UPDATE` or `DELETE` statements. So, this mostly affects the `UpdatableRecordImpl::storeMergeOrUpdate0` branch of the call tree, in case the solution is ignoring the key.

There's another thought. The whole logic where we add these conditions might be flawed in PostgreSQL, as the `CASE` statement only runs checks on the primary key, not the unique key as well. I'll investigate this further, possibly in a separate issue.

**Comment by lukaseder** [comment]
Removing the conditions for the `merge()` case only doesn't have any negative impacts in MySQL (`ON DUPLICATE KEY UPDATE`) or H2 (`MERGE`) implementations. In PostgreSQL, the query is wrong either way because we can't list more than 1 constraint in the `ON CONFLICT` clause, see https://github.com/jOOQ/jOOQ/issues/7552.

There, the generated SQL is:

```sql
insert into t_11552 (CD, XY)
values (
  1, 
  2
)
on conflict (ID)
do update
set
  CD = 1,
  XY = 2
where t_11552.ID is null
returning t_11552.ID
```

A thing to be considered here that would fix both the issue you reported and the above PostgreSQL problem is that we iterate the unique keys until we find the first one in the record that has its values `changed()`

**Comment by lukaseder** [comment]
I won't rush addressing the PostgreSQL case yet, see discussion here: https://github.com/jOOQ/jOOQ/issues/7552#issuecomment-1064113176

It's already great if this bug here can be fixed for all non-PG dialects that support MySQL's or the standard's variant of the syntax.

**Comment by lukaseder** [comment]
Implemented in jOOQ 3.17.0, 3.16.6 (https://github.com/jOOQ/jOOQ/issues/13256), and 3.15.10 (https://github.com/jOOQ/jOOQ/issues/13257). Thanks again for your report!

**Comment by lukaseder** [comment]
I'll add a few more integration tests for `UpdatableRecord` operations on that `T_11552` table. There might be other issues lurking.
